File: backend/app/routes/coding.py
"""Coding statistics endpoints."""

from __future__ import annotations

from fastapi import APIRouter, HTTPException, Query

from ..schemas import CodingPlatformStats, CodingSummary
from ..services import coding_service

router = APIRouter(prefix="/coding", tags=["coding"])

_VALID = {"github", "leetcode", "codechef", "hackerrank", "geeksforgeeks", "codeforces"}


@router.get("", response_model=CodingSummary)
async def coding_summary(force: bool = Query(False, description="Bypass cache")) -> CodingSummary:
    return await coding_service.get_summary(force=force)


@router.get("/{platform}", response_model=CodingPlatformStats)
async def coding_platform(
    platform: str, force: bool = Query(False, description="Bypass cache")
) -> CodingPlatformStats:
    if platform not in _VALID:
        raise HTTPException(status_code=404, detail=f"Unknown platform '{platform}'.")
    return await coding_service.get_platform(platform, force=force)

File: backend/src/routes/coding.ts
/** Coding statistics endpoints. */

import { Router } from 'express'
import * as coding from '../services/codingService'

export const codingRouter = Router()

const VALID = new Set(['github', 'leetcode', 'codechef', 'hackerrank', 'geeksforgeeks', 'codeforces'])
const HEATMAP_VALID = new Set(['github', 'leetcode'])

const wantsForce = (v: unknown): boolean => v === 'true' || v === '1'

codingRouter.get('/coding', async (req, res, next) => {
  try {
    res.json(await coding.getSummary(wantsForce(req.query.force)))
  } catch (err) {
    next(err)
  }
})

codingRouter.get('/coding/:platform/heatmap', async (req, res, next) => {
  try {
    const { platform } = req.params
    if (!HEATMAP_VALID.has(platform)) {
      res.status(404).json({ detail: `No activity heatmap for platform '${platform}'.` })
      return
    }
    res.json(await coding.getHeatmap(platform, wantsForce(req.query.force)))
  } catch (err) {
    next(err)
  }
})

codingRouter.get('/coding/:platform', async (req, res, next) => {
  try {
    const { platform } = req.params
    if (!VALID.has(platform)) {
      res.status(404).json({ detail: `Unknown platform '${platform}'.` })
      return
    }
    res.json(await coding.getPlatform(platform, wantsForce(req.query.force)))
  } catch (err) {
    next(err)
  }
})
